Add index to ATen XLA tensor

asuhan opened the pull request.

dlibenzi commented on torch_xla/csrc/index_op_util.cpp:
```diff
@@ -0,0 +1,103 @@
+            << self.sizes() << " at index " << src_idx;
+      }
+      // Replace with nonzeros.
+      auto nonzero = index.nonzero();
```
If the majority of use cases pass in Byte tensors, we did all this for nothing (or almost so) :|

asuhan replied:
That's a distinct possibility, could still be a win if we just do the `nonzero` on the host and finish the gather on device. But yeah, not great.

asuhan commented on torch_xla/csrc/index_op_util.cpp:
```diff
@@ -0,0 +1,104 @@
+#include "torch_xla/csrc/index_op_util.h"
```
The main reason for this is avoiding weird dependencies. We'd have something under `ops` depend on `at::Tensor` and `XLATensor`. Would you rather have that instead of the current situation? That was the main reason I split them.

dlibenzi replied:
There is no weird dependency AFAICT.
The index_op.cpp/h contains all the infrstructure to do this thing, takes input ir::Value/Node and produces IR Node.

asuhan replied:
There is because of the canonicalization of the null bits, if we put everything in one place.

asuhan replied:
Done.

asuhan commented on torch_xla/csrc/aten_xla_type.cpp:
```diff
@@ -762,6 +763,15 @@ at::Tensor AtenXlaType::expand_as(const at::Tensor& self,
+                              at::TensorList indices) const {
+  CanonicalIndexInfo canonical_index_info =
+      GetCanonicalIndexInfo(self, indices);
+  return bridge::AtenFromXlaTensor(
```
You mean as `at::Tensor`s? That's a bit of weird, mixing `XLATensor` with `at::Tensor`.

asuhan commented on torch_xla/csrc/tensor.cpp:
```diff
@@ -1835,6 +1847,24 @@ Device XLATensor::CommonDeviceForTensors(
   return device;
 }
 
+std::vector<XLATensor> XLATensor::WrapIndicesOnce(
```
One of the reasons it's not there is because `XLATensor::Create` is private, which is probably a good thing to preserve. I could take the contents of the `XLATensor` methods returning tensor(s) which are used by `XLATensor::WrapIndicesOnce` and `XLATensor::index`, but having to do that rather suggests this method belongs to `XLATensor`.

dlibenzi replied:
The index() will be in tensor.cpp .. the creation of the IR Node, in index_op.cpp/h
As well as all its handling.

asuhan replied:
Yes, we can do that at the cost of inlining implementations of `lt` and `where`. They are tiny, it's fine, I'll do it that way.

dlibenzi replied:
Let me re-check the code ... I am not sure what you mean for inlining ...

asuhan replied:
Done.

dlibenzi commented on torch_xla/csrc/tensor.cpp:
```diff
@@ -1835,6 +1847,24 @@ Device XLATensor::CommonDeviceForTensors(
+        Create(dim_index.GetIrValue() +
+                   ir::ops::ScalarOp(at::Scalar(dim_size), dim_index.shape()),
+               input.GetDevice());
+    XLATensor wrap_cond = lt(indices[dim_idx], at::Scalar(int64_t(0)));
```
You can just call XLATensor::lt() and XLATensor::where() ... they are static anyway.

asuhan replied:
Ok, but they take tensors. So we'd have a situation in which we pass `XLATensor` inputs and get back an `ir::Value`, then? That also means the `IndexOp` would be subject to the same weirdness of receiving `XLATensor` inputs.

dlibenzi replied:
I don't understand. You are handling XLATensors here ... and the XLATensor::lt() and XLATensor::where() APIs are both public and static.
What is the problem?
You mean that XLATensor will use inde_op.cpp/h, and vice versa?
As long as include files don't cross-fire to each other, it should be OK.
Not pretty, but OK.

asuhan replied:
Ok, let's recap. If we move `WrapIndicesOnce` out of `XLATensor`, we lose access to `XLATensor::Create` (since it's private). This means we can't create `wrapped_dim_index` anymore as an `XLATensor`, which in turn prevents us from using `where`.

dlibenzi replied:
I think creating an XLATensor from an IR Node, can be made public. I do not see big damage there.
Alternatively, create an at::Tensor from Scalar+sizes, and use the public ctor.

asuhan replied:
Alright, I'll make that `Create` overload public and be done with it.

dlibenzi replied:
I think it might get useful in other cases where we have some infrastructure that do not strictly belong to XLATensor, but still wanting to create them based on IR Nodes.

asuhan replied:
Sounds good, done.

asuhan commented on torch_xla/csrc/index_op_util.cpp:
```diff
@@ -0,0 +1,104 @@
+      auto& type = tensor.type();
+      auto scalarType = type.scalarType();
+      if (scalarType != at::kLong && scalarType != at::kByte) {
+        XLA_ERROR() << "tensors used as indices must be long or byte tensors";
```
Done.

asuhan commented on torch_xla/csrc/index_op_util.cpp:
```diff
@@ -0,0 +1,104 @@
+  for (auto& tensor : indices) {
+    if (tensor.defined()) {
+      auto& type = tensor.type();
+      auto scalarType = type.scalarType();
```
Done.

asuhan commented on torch_xla/csrc/tensor.h:
```diff
@@ -695,6 +702,12 @@ class XLATensor {
 
+  // Wraps index tensors once into the [0, dim_size) interval, where dim_size is
+  // the size of the current indexed dimension.
+  static std::vector<XLATensor> WrapIndicesOnce(
```
Done.

asuhan commented on torch_xla/csrc/ops/ops.cpp:
```diff
@@ -401,6 +401,22 @@ NodePtr BroadcastTensors(tensorflow::gtl::ArraySlice<const Value> tensors) {
                    std::move(lower_fn), /*num_outputs=*/tensors.size());
 }
 
+NodePtr IndexOp(const Value& input, const Value& indices) {
```
Done.

asuhan commented:
I've addressed all but one comment, please take a look.

dlibenzi commented on torch_xla/csrc/ops/index_op.cpp:
```diff
@@ -0,0 +1,163 @@
+  for (auto& tensor : indices) {
+    if (tensor.defined()) {
+      auto& type = tensor.type();
+      auto scalar_type = type.scalarType();
```
This is not a template, is it? So maybe at::ScalarType ...

asuhan replied:
Done.

dlibenzi commented on torch_xla/csrc/ops/index_op.cpp:
```diff
@@ -0,0 +1,163 @@
+#include "torch_xla/csrc/ops/index_op.h"
```
I think we leave an empty line by style:

```
#include "THIS_CPP_H"

#include <SYSTEM_INCS>
...

#include <NON_LOCAL_INCS>
...

#include "LOCAL_INCS"
...

namespace ... {
```

asuhan replied:
Done.

dlibenzi commented on torch_xla/csrc/ops/index_op.cpp:
```diff
@@ -0,0 +1,163 @@
+  size_t base_rank = base.dim();
+  dims.reserve(base_rank);
+  XLA_CHECK_LE(indices.size(), base_rank);
+  for (size_t i = 0; i < indices.size(); i++) {
```
Care when we do loops like there on calls the compiler cannot see through, as it keep calling the bound API for every iteration:

https://gcc.godbolt.org/z/BG5HlI

asuhan replied:
Realistically, we can only have maybe a dozen of indices.

dlibenzi replied:
Sure, but the thing can keep creating shapes for some APIs.
The one above, is not issue. But things which end up calling into tensor methods, they might be.

asuhan replied:
Ok, I don't understand the resolution here, neither the comment about shapes. If it's about dynamic models, the shape sizes will change, not the rank. We're not going to have rank 100k anytime soon.

dlibenzi replied:
It's OK here. Mine was only a note ;)

dlibenzi commented on torch_xla/csrc/ops/index_op.cpp:
```diff
@@ -0,0 +1,163 @@
+    const XLATensor& base,
+    tensorflow::gtl::ArraySlice<const XLATensor> indices) {
+  std::vector<XLATensor> canonical_indices;
+  XLA_CHECK_LE(indices.size(), base.shape().get().rank());
```
Cache base.shape() as we are using it below.

asuhan replied:
Done.

asuhan commented on torch_xla/csrc/ops/index_op.cpp:
```diff
@@ -0,0 +1,172 @@
+    if (tensor.defined()) {
+      at::ScalarType scalar_type = tensor.scalar_type();
+      if (scalar_type != at::kLong &&
+          scalar_type != XlaTensorTypeMapping(at::kByte, device)) {
```
@dlibenzi Because TPU is using `S32` to map `Byte`, we have to do this ugly thing. Let me know if you see a better way.

dlibenzi replied:
Hold on one second ...

dlibenzi replied:
How about this?

https://github.com/pytorch/xla/pull/356

asuhan replied:
Yeah, probably a better option.

asuhan replied:
Unfortunately it didn't help, still getting `kInt` here even with your patch applied. I'm looking into it more.

dlibenzi replied:
Then it is worse, and I think I know what is going on.
When we manipulate tensors, we lose that info:

```
Byte a = 1;
Byte b = 2;
auto c = a + b;

type(c) == Int
```

I am not sure the correct solution is to propagate the logic type.
Even because that is a HUGE change.

asuhan replied:
We need to preserve the logical type across calls into the bridge. Shouldn't `bridge::CreateXlaTensor` take the logical type?

dlibenzi replied:
The XLATensor::Create(ir::Value, Device) should take c10::optional<at::ScalarType>.

dlibenzi replied:
Let me think for a second before dropping int code ...

asuhan replied:
Maybe for your particular example, for this indexing problem looks like `Create(const at::Tensor& tensor, const Device& device, bool requires_grad)` is the culprit.

dlibenzi replied:
Essentially, for the majority of the cases besides "cast" APIs, we'd need something like:

static XLATensor CreateFrom(const XLATensor& input, ir::Value value);

Or the equivalent:

XLATensor CreateFrom(ir::Value value) const;

The API would then call the Create(ir::Value, Device, optional<ScalarType>);

dlibenzi replied:
Let me sketch one in a diff, then you can do it.
Or I do it.
We could decouple this, if the change required by the index PR to deal with it, is not huge.

But if we have code that wants to infer types, like the index() one, we need to propagate them.
Hopefully this won't come back bite our arse later.

dlibenzi replied:
To give the idea:

```
diff --git a/torch_xla/csrc/tensor.cpp b/torch_xla/csrc/tensor.cpp
index 6495d79..89fbb8a 100644
--- a/torch_xla/csrc/tensor.cpp
+++ b/torch_xla/csrc/tensor.cpp
@@ -538,8 +538,7 @@ XLATensor::ViewIrNode XLATensor::GetViewIrNode(View* view) {
 XLATensor XLATensor::add(const XLATensor& input, const XLATensor& other,
                          const at::Scalar& alpha) {
   ir::NodePtr constant = ir::ops::ScalarOp(alpha, other.shape());
-  return Create(input.GetIrValue() + other.GetIrValue() * constant,
-                input.GetDevice());
+  return input.CreateFrom(input.GetIrValue() + other.GetIrValue() * constant);
 }
 
 void XLATensor::add_(XLATensor& input, const XLATensor& other,
@@ -551,8 +550,7 @@ void XLATensor::add_(XLATensor& input, const XLATensor& other,
 XLATensor XLATensor::sub(const XLATensor& input, const XLATensor& other,
                          const at::Scalar& alpha) {
   ir::NodePtr constant = ir::ops::ScalarOp(alpha, other.shape());
-  return Create(input.GetIrValue() - other.GetIrValue() * constant,
-                input.GetDevice());
+  return input.CreateFrom(input.GetIrValue() - other.GetIrValue() * constant);
 }
 
 void XLATensor::sub_(XLATensor& input, const XLATensor& other,
@@ -1487,6 +1485,10 @@ XLATensor XLATensor::CreateView(
   return Create(std::make_shared<View>(std::move(shape), alias), GetDevice());
 }
 
+XLATensor XLATensor::CreateFrom(ir::Value ir_value) const {
+  return Create(std::move(ir_value), GetDevice(), dtype());
+}
+
 XLATensor XLATensor::log_softmax(const XLATensor& input, xla::int64 dim) {
   return Create(ir::MakeNode<ir::ops::LogSoftmax>(input.GetIrValue(), dim),
                 input.GetDevice());
diff --git a/torch_xla/csrc/tensor.h b/torch_xla/csrc/tensor.h
index 6a27f47..2891603 100644
--- a/torch_xla/csrc/tensor.h
+++ b/torch_xla/csrc/tensor.h
@@ -646,6 +646,10 @@ class XLATensor {
   XLATensor CreateView(
       tensorflow::gtl::ArraySlice<const xla::int64> output_size) const;
 
+  // Create a new XLA tensor with the same metadata of the input tensor, and the
+  // new IR value.
+  XLATensor CreateFrom(ir::Value ir_value) const;
+
   // Discards all the XLA and IR data, by making the ATEN tensor one the only
   // source for this XLA tensor. An error is generated if the XLA tensor does
   // not have ATEN tensors data.
```

dlibenzi replied:
That way, if we ever have to change the metadata to be propagated, we have a single funnel point.

asuhan replied:
To land this PR, it is sufficient to set the logical type for this constructor: https://github.com/pytorch/xla/blob/master/torch_xla/csrc/tensor.h#L600. Anything wrong about doing that? It looks like a bug fix to me.

dlibenzi replied:
That makes sense indeed.
If that fixes your PR, good!
Let's go ahead with it.
I will take care of the other thing, because once we do operations of tensors, we lose the information.

asuhan replied:
Ok, I'll do that and land this.

Regarding your sketch, it does make sense. As long as we have IR nodes at all sites which require doing this, it'll be fine. Basically, we need two things:

1. When ATen tensors enter our system, preserve their logical type. This is achieved by fixing https://github.com/pytorch/xla/blob/master/torch_xla/csrc/tensor.h#L600.
2. We have to preserve the types across operations, which is achieved through your proposal.

By virtue of structural induction, we should be fine in all cases.

asuhan commented:
I've removed the type mapping commit, rebased on top of master and added https://github.com/pytorch/xla/pull/350/commits/9c811b4205acf47799d6bfd3fa0ae630569fe83e to fix the conversion from byte tensor issue. Please take a look.
